### ISSUE TITLE
[iCalendar] Provide a (configurable) amount of upcoming events

### ISSUE THREAD

**Issue by Rodelkoenig** [openhab/openhab-addons#8022]
## Current Situation
This binding currently provides a channel for a currently running event and a channel for the next upcoming event.

## Goal
I, as a user, want to implement the calendar for my garbage collection. There are up to 3 events/collections at one day and I want to configure the binding to display a fixed amount of upcoming events. These events are used for display purposes and used in rules to inform me for upcoming garbage collection the next day.

## My idea
I am currently using the v1-caldav-personal-addon for the use-case above which works fine for me.
Is it possible to set
1) a fixed number of events in Thing-configuration OR
1) a fixed number of minutes/hours/days in Thing-configuration to get all events in this frame

and a Channel which contains a list of events which are determined by the parameter above.

### Disclaimer
I am a new openHAB user and I am currently not firm with the basic structure of V2-Addons/Bindings. If there is a better idea to implement my idea than my suggestion please tell me.

I am also a Java developer so I could eventually contribute to the development if it is okay for you.

**Comment by daMihe** [comment]
AFAIK: It is not possible this way as the things channels have to be static for a type.

However, we had the idea to move icalendar from a single thing per calendar to a bridge. A bridge behaves same but has things itself. These "subthings" of same type may behave differently based on parameters but have all the same structure - this is the point where to start in my eyes.

@andrewfg What do you think about following todos?
* Move calendar thing type to a bridge and keep the behavior as is
* Add thing type to the bridge which implements filters (maybe first a simple text filter for just showing matching events)

**Comment by jeremystucki** [comment]
Wouldn't it be possible to create dynamic channels for a configurable number of events? Like I did for the feed binding in PR #8070

**Comment by andrewfg** [comment]
> AFAIK: It is not possible this way as the things channels have to be static for a type.

@daMihe actually it is perfectly ok to have dynamic channels; there are several bindings (Velux, PowerView, to name but two) which create channels dynamically for “scenes”; I can share some code with you for that. However, I think it is important to have some config parameters that would limit the quantity of dynamic channels, since it otherwise risks killing the system.

**Comment by andrewfg** [comment]
Following is some code (from the HD PowerView binding) that adds a "scene" Channel to a Thing dynamically. Note: this is the code for **adding** a Channel. However in the case of iCalendar, you would obviously also need code for **deleting** any already created Channels which are no longer valid, to prevent a "Channel/memory leak".

```
// create a new scene channel
ChannelUID channelUID = new ChannelUID(getThing().getUID(), sceneId);
Channel channel = ChannelBuilder.create(channelUID, "Switch")
    .withType(sceneChannelTypeUID)
    .withLabel(scene.getName())
    .withDescription("Activates the scene " + scene.getName())
    .build();
updateThing(editThing().withChannel(channel).build());
```

**Comment by daMihe** [comment]
I'll look into that approach next time. If you want, you can assign me to the issue.

**Comment by daMihe** [comment]
Progress! If you want , you can take a look into https://github.com/daMihe/openhab-addons/tree/8022-icalendar-filtered-events. Currently the code is far from perfect but the algorithm itself works and events are shown correctly.

**Comment by andrewfg** [comment]
Hi @daMihe many thanks for the update.

I see what you have added to the code, but I don't really see what is the use case, so I can't really comment on it or test the functionality much -- perhaps you should describe the update in `ReadMe.md` first ?? :)

My first observation is that you seem to have converted the iCalendar object from being a Thing to being a Bridge (??). And I don't really understand why you did that. Could you not just add Channels (dynamically created for the filtered events) to your existing iCalendar Thing object, without changing its type from Thing to Bridge. Or am I overlooking something fundamental?

I worry that if you change the object from Thing to Bridge it would be a "breaking change" for existing users. (??)

**Comment by daMihe** [comment]
The state of the code (including documentation) is why there is currently no pull request. I'll improve it before doing the PR. The reason for changing the thing type to a bridge is that it allows multiple filters per calendar. This way, only one calendar is kept and is the source of truth. The "Filtered Events"-ThingType just allows one (let's name it) query based on a combination of time and text. Doing multiple queries would be harder to implement in one monolithic calendar thing.

As the id's did not change and bridges are also items the transition should not be a problem. But testing the behaviour is needed.

The frontend (PaperUI) currently documents the possibilities quite good. Sure, some descriptions should be added to the readme - as my english is bad, i'm keeping this as TODO until the texts are semi final.

**Comment by daMihe** [comment]
@andrewfg You were right installing over old version causes trouble. I've now implemented a workaround, which recreates the thing before creating the handler. A notice in readme should now be enough.

**Comment by tfelix** [comment]
Maybe to add something to the thought: I have a similar issue, i have a whole-day event like e.g holydays and if I have a short event it will only the event which "overlays" the holyday event. See the attached picture. After 10o'clock only "Test" is shown as the active element but not the "Urlaub". Maybe its possible to just concat the event names into a single channel e.g. something like "Urlaub\nTest" shows up, a bit sketchy though? Or add something like a filter. I think the older bindings for e.g. caldav had such a functionality. I did not yet have time to look into the code, how its done there.

![image](https://user-images.githubusercontent.com/1186998/92667872-73bb4a80-f30d-11ea-8a74-7c09fe9e6537.png)

**Comment by politas** [comment]
Here's another use case:  

- I'm a shift worker
- If I have an upcoming day shift tomorrow morning, I set a status to adjust how heating and lighting is handled in the hours leading up to the shift, to maximise my ability to sleep well.
- If I have an upcoming night shift starting this evening, likewise.
- If I have an upcoming calendar event during the night or day before one of those shifts, my rules looking to my next event cannot cannot see the work shift, so those status items cannot be set correctly.

**Comment by daMihe** [comment]
@tfelix i don't like this idea. It is causing intransparencies if you want to handle events automatically. The current event title also should change back to "Urlaub" after "Test" is over. Concatenating is not everything in this case. It would be also required to define a specific behavior for the begin and end datetimes which is much harder to do in a way which is logical.

@politas This should be now (including this PR) possible if you choose the right parameters for the time frame.

**Comment by politas** [comment]
> @politas This should be now (including this PR) possible if you choose the right parameters for the time frame.

That sounds like exactly what I want!

**Comment by tfelix** [comment]
> @tfelix i don't like this idea. It is causing intransparencies if you want to handle events automatically. The current event title also should change back to "Urlaub" after "Test" is over. Concatenating is not everything in this case. It would be also required to define a specific behavior for the begin and end datetimes which is much harder to do in a way which is logical.

Fair enough, actually I can just define a rule and react upon the daylong event as soon as it pops up in the item as I only need to trigger that once. So I guess I can work around that. Thanks for the reply and the idea.

**Comment by daMihe** [comment]
@andrewfg could you review my changes (especially the README and texts)?

**Comment by andrewfg** [comment]
>  could you review my changes (especially the README and texts)?

Done.

**Comment by daMihe** [comment]
Thank you. The subheaders were added as suggested.

**Comment by daMihe** [comment]
@tfelix, @politas: As build was successful at the OpenHAB build infrastructure you can download the addon directly from artifactory: https://openhab.jfrog.io/openhab/webapp/#/builds/PR-openHAB-Addons/285/1601151137665/published%20/org.openhab.addons.bundles:org.openhab.binding.icalendar:3.0.0-SNAPSHOT 

Note the code was built for OpenHAB 3.

**Comment by politas** [comment]
> @tfelix, @politas: As build was successful at the OpenHAB build infrastructure you can download the addon directly from artifactory: https://openhab.jfrog.io/openhab/webapp/#/builds/PR-openHAB-Addons/285/1601151137665/published%20/org.openhab.addons.bundles:org.openhab.binding.icalendar:3.0.0-SNAPSHOT 

Ok, that'll be a new adventure! 

> Note the code was built for OpenHAB 3.

Does that mean it won't work on OpenHAB 2.x?

**Comment by daMihe** [comment]
No, as the OpenHAB core developers decided to switch the main development branch to OH 3 i don't see any chance a PR for 2.5.x will still be accepted. I think even 2.5.10 won't be released, 3.0 will be instead. Please check your addons - legacy addons won't work.
